File: src/workers/chapter-worker.ts
import { eq } from 'drizzle-orm';
// import { Chapter, chapters, NewChapter } from '';
import { Chapter, chapters, db, NewChapter } from '../db/index';

export class ChapterWorker {
  public async createChapter(data: NewChapter): Promise<Chapter[]> {
    return await db.insert(chapters).values(data).returning();
  }

  public async getChaptersByMangaId(mangaId: number): Promise<Chapter[]> {
    return await db.select().from(chapters).where(eq(chapters.mangaId, mangaId)).orderBy(chapters.chapterNumber);
  }

  public async getChaptersByVolumeId(volumeId: number): Promise<Chapter[]> {
    return await db.select().from(chapters).where(eq(chapters.volumeId, volumeId)).orderBy(chapters.chapterNumber);
  }

  public async updateChapter(id: number, data: Partial<Omit<NewChapter, 'id'>>): Promise<Chapter | undefined> {
    const [updatedChapter] = await db
      .update(chapters)
      .set({ ...data, updatedAt: new Date() })
      .where(eq(chapters.id, id))
      .returning();
    return updatedChapter;
  }

  public async markChapterAsDownloaded(id: number): Promise<Chapter | undefined> {
    const [updatedChapter] = await db.update(chapters).set({ downloaded: true, updatedAt: new Date() }).where(eq(chapters.id, id)).returning();
    return updatedChapter;
  }

  public async updateChapterUrl(chapterId: number, chapterUrl: string): Promise<Chapter | undefined> {
    return await this.updateChapter(chapterId, { url: chapterUrl });
  }
}

File: Service/DataWorkers/ChapterWorker.cs
using System.Data;
using Dapper;
public class ChapterWorker : IChapterWorker
{
    private readonly IDbConnection _connection;
    public ChapterWorker(IDbConnection connection)
    {
        _connection = connection;
    }

    public async Task<IEnumerable<Chapter>?> GetChaptersByManga(long id)
    {
        var response = await _connection.QueryAsync<Chapter>(SqlScripts.Chapters.GetByMangaId, id);

        return response;
    }

    public async Task<IEnumerable<Chapter>?> GetChaptersByVolume(long id)
    {
        var response = await _connection.QueryAsync<Chapter>(SqlScripts.Chapters.GetByVolumeId, id);

        return response;
    }

    public async Task<Volume?> CreateChapter(ChapterRequest data)
    {
        var response = (await _connection.QueryAsync<Volume>(SqlScripts.Chapters.Insert, new
        {
            data.MangaId,
            data.VolumeId,
            data.ChapterKeyword,
            data.ChapterNumber,
            data.Title,
            data.Url,
            data.Downloaded,
            CreatedAt = DateTime.Now,
            UpdatedAt = DateTime.Now
        })).FirstOrDefault();

        return response;
    }

    public async Task<Volume?> UpdateChapter(ChapterUpdateRequest data)
    {
        var response = (await _connection.QueryAsync<Volume>(SqlScripts.Chapters.Update, new
        {
            data.Id,
            data.ChapterKeyword,
            data.ChapterNumber,
            data.Title,
            data.Url,
            data.Downloaded,
            CreatedAt = DateTime.Now,
            UpdatedAt = DateTime.Now
        })).FirstOrDefault();

        return response;
    }

    public async Task<long> DeleteChapter(long id)
    {
        return (await _connection.QueryAsync<long>(SqlScripts.Chapters.Delete, id)).FirstOrDefault();
    }



}
